pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -259,6 +261,20 @@ func (configgen *ConfigGeneratorImpl) buildGatewayHTTPRouteConfig(node *model.Pr
 				continue
 			}
 
+			// Make sure we can obtain services which are visible to this virtualService.
+			nameToServiceMap, exist := namespaceAndHostname[virtualService.Namespace]
+			if !exist {
+				services := push.Services(&model.Proxy{
+					ConfigNamespace: virtualService.Namespace,
```

Comment:
You should not assume the vs' namespace == gateway's namespace

Reply:
When pilot build RDS for gateway, it will select servers of gateways based on route name generated by LDS. Then it will select the `virtualservices` based on the namespace of proxy node and gateway name of server selected by first step. So here, it means that the selected `virtualservices` are exported to gateway. The hostname of service is directly referred by VirtualService not Gateway. So we only need make sure the services are visible to VirtualService. 

Please correct me if I am wrong.